### PR TITLE
Optimize MiniCPMO mask creation with vectorized implementation

### PR DESCRIPTION
**Pull request by skyloevil**
# MiniCPMO Mask Creation Optimization Summary

## Optimization Overview

This PR optimizes the `subsequent_chunk_mask` function in MiniCPMO by replacing loop-based mask generation with vectorized tensor operations, achieving **2.96x average speedup** across different input sizes.

### Test Results Summary
- **Testing Environment**: MacBook Pro M4, 48GB RAM
- **Correctness**: All test cases pass - identical output to original implementation
- **Average Speedup**: **2.96x** (up to **5.49x** for certain sizes)
- **Memory Usage**: No additional memory overhead
- **Compatibility**: 100% drop-in replacement

### Detailed Performance Comparison

| Size | Original (ms) | Optimized (ms) | Speedup |
|------|---------------|----------------|---------|
| 64   | 0.404         | 0.279          | 1.45x   |
| 128  | 0.184         | 0.034          | **5.49x** |
| 256  | 0.566         | 0.163          | 3.48x   |
| 512  | 0.825         | 0.312          | 2.64x   |
| 1024 | 1.733         | 0.593          | 2.92x   |
| 2048 | 3.239         | 1.841          | 1.76x   |

### Before (Loop-based Implementation)
```python
def subsequent_chunk_mask(self, size: int, chunk_size: int, num_left_chunks: int = -1, 
                         device: torch.device = CPU_DEVICE, num_lookhead: int = 0) -> torch.Tensor:
    ret = torch.zeros(size, size, device=device, dtype=torch.bool)
    for i in range(size):  # Slow row-by-row processing
        if num_left_chunks < 0:
            start = 0
        else:
            start = max((i // chunk_size - num_left_chunks) * chunk_size, 0)
        ending = min((i // chunk_size + 1) * chunk_size + num_lookhead, size)
        ret[i, start:ending] = True
    return ret
```

### After (Vectorized Implementation)
```python
def subsequent_chunk_mask(self, size: int, chunk_size: int, num_left_chunks: int = -1,
                         device: torch.device = CPU_DEVICE, num_lookhead: int = 0) -> torch.Tensor:
    ret = torch.zeros(size, size, device=device, dtype=torch.bool)
    
    # Vectorized computation of row indices and chunk boundaries
    row_indices = torch.arange(size, device=device)
    chunk_indices = row_indices // chunk_size
    
    if num_left_chunks < 0:
        start_indices = torch.zeros_like(row_indices)
    else:
        start_chunk_indices = torch.clamp(chunk_indices - num_left_chunks, min=0)
        start_indices = start_chunk_indices * chunk_size
    
    # Compute ending indices vectorially
    end_chunk_indices = chunk_indices + 1
    end_indices = torch.clamp(end_chunk_indices * chunk_size + num_lookhead, max=size)
    
    # Create column indices for broadcasting
    col_indices = torch.arange(size, device=device).unsqueeze(0)
    row_indices = row_indices.unsqueeze(1)
    start_indices = start_indices.unsqueeze(1)
    end_indices = end_indices.unsqueeze(1)
    
    # Vectorized mask creation using broadcasting
    ret = (col_indices >= start_indices) & (col_indices < end_indices)
    
    return ret
```

## Key Optimization Techniques

1. **Vectorized Index Computation**: Replaced `for i in range(size)` with `torch.arange(size)`
2. **Batch Boundary Calculation**: Computed start/end indices for all rows simultaneously
3. **Broadcasting Magic**: Used tensor broadcasting to create the entire mask in one operation
4. **Efficient Tensor Operations**: Leveraged `torch.clamp` for boundary constraints

### Original Approach Problems:
- **Sequential Processing**: Processes one row at a time
- **Python Loop Overhead**: Each iteration has Python call overhead
- **Memory Access Pattern**: Not optimized for cache locality

### Optimized Approach Benefits:
- **Parallel Processing**: All computations happen simultaneously
- **GPU/SIMD Friendly**: Vectorized operations utilize hardware acceleration
- **Cache Efficient**: Better memory access patterns
- **No Python Overhead**: Pure tensor operations

## Example Mask Visualization

For `size=16, chunk_size=4, num_left_chunks=1, num_lookhead=2`:

```
     0  1  2  3  4  5  6  7  8  9 10 11 12 13 14 15
 0:  1  1  1  1  1  1  0  0  0  0  0  0  0  0  0  0
 1:  1  1  1  1  1  1  0  0  0  0  0  0  0  0  0  0
 2:  1  1  1  1  1  1  0  0  0  0  0  0  0  0  0  0
 3:  1  1  1  1  1  1  0  0  0  0  0  0  0  0  0  0
 4:  1  1  1  1  1  1  1  1  1  1  0  0  0  0  0  0
 5:  1  1  1  1  1  1  1  1  1  1  0  0  0  0  0  0
 6:  1  1  1  1  1  1  1  1  1  1  0  0  0  0  0  0
 7:  1  1  1  1  1  1  1  1  1  1  0  0  0  0  0  0
 8:  0  0  0  0  1  1  1  1  1  1  1  1  1  1  0  0
 9:  0  0  0  0  1  1  1  1  1  1  1  1  1  1  0  0
10:  0  0  0  0  1  1  1  1  1  1  1  1  1  1  0  0
11:  0  0  0  0  1  1  1  1  1  1  1  1  1  1  0  0
12:  0  0  0  0  0  0  0  0  1  1  1  1  1  1  1  1
13:  0  0  0  0  0  0  0  0  1  1  1  1  1  1  1  1
14:  0  0  0  0  0  0  0  0  1  1  1  1  1  1  1  1
15:  0  0  0  0  0  0  0  0  1  1  1  1  1  1  1  1
```

## Testing & Validation

- **Correctness Tests**: 5 different parameter combinations - all passed
- **Performance Tests**: Tested on sizes from 64 to 2048
- **Memory Tests**: No additional memory overhead confirmed
- **Edge Cases**: Tested various chunk sizes and lookahead values

## Impact

This optimization provides significant performance improvements for MiniCPMO's attention mask creation, especially benefiting:

- **Training Speed**: Faster mask generation during training
- **Inference Performance**: Reduced latency for attention computation
- **Memory Efficiency**: Same memory footprint with better performance
- **Scalability**: Better performance scaling with larger sequence lengths

The optimization maintains 100% compatibility while delivering substantial performance gains through modern PyTorch vectorization techniques.

### CODE REVIEW

**Comment by DarkLight1337** [comment]
cc @tc-mb can you verify this?

**Comment by tc-mb** [comment]
> cc @tc-mb can you verify this?

Of course, I will look into verifying the optimization and helping with this PR.

**Comment by tc-mb** [comment]
@DarkLight1337  I ran a side-by-side test, and the results are correct. This is a good optimization.

@skyloevil  Could I incorporate this optimization into other code implementations? I'll include a link to this pull request and your account name as a reference.

**Review by DarkLight1337 — APPROVED** [review]

**Comment by skyloevil** [comment]
> @DarkLight1337 I ran a side-by-side test, and the results are correct. This is a good optimization.
> 
> @skyloevil Could I incorporate this optimization into other code implementations? I'll include a link to this pull request and your account name as a reference.

Hi @tc-mb,Great to see this optimization being useful elsewhere! Approved with attribution - go ahead and propagate it to other projects 🚀

**Comment by tc-mb** [comment]
> > @DarkLight1337 I ran a side-by-side test, and the results are correct. This is a good optimization.
> > @skyloevil Could I incorporate this optimization into other code implementations? I'll include a link to this pull request and your account name as a reference.
> 
> Hi @tc-mb,Great to see this optimization being useful elsewhere! Approved with attribution - go ahead and propagate it to other projects 🚀

thx ^_^
